ledger.go

```diff
@@ -380,56 +383,186 @@ func (l *Ledger) PullTransactions() {
 			select {
 			case <-l.sync:
 				return
-			case <-time.After(1 * time.Second):
+			default:
 			}
 
 			continue
 		}
 
-		logger := log.Sync("pull_tx")
+		logger := log.Sync("sync_tx")
 
-		// Build list of transaction IDs
-		req := &TransactionPullRequest{
-			TransactionIds: make([][]byte, 0, l.transactions.Len()),
+		buf := bytes.NewBuffer(nil)
+		if _, err := l.transactionsSyncIndex.WriteTo(buf); err != nil {
+			logger.Error().Err(err).Msg("failed to marshal bloom filter")
+			continue
 		}
 
-		l.transactions.Iterate(func(tx *Transaction) bool {
-			req.TransactionIds = append(req.TransactionIds, tx.ID[:])
-			return true
-		})
+		var wg sync.WaitGroup
 
-		workerChan := make(chan *grpc.ClientConn, 16)
-		var workerWG sync.WaitGroup
-		workerWG.Add(cap(workerChan))
-		responseChan := make(chan *TransactionPullResponse, snowballK)
-		for i := 0; i < cap(workerChan); i++ {
-			go func() {
-				for conn := range workerChan {
-					client := NewWaveletClient(conn)
+		bfReq := &TransactionsSyncRequest{
+			Data: &TransactionsSyncRequest_Filter{
+				Filter: buf.Bytes(),
+			},
+		}
 
-					ctx, cancel := context.WithTimeout(context.Background(), conf.GetDownloadTxTimeout())
-					batch, err := client.PullTransactions(ctx, req)
-					if err != nil {
-						logger.Error().Err(err).Msg("failed to download missing transactions")
-						cancel()
+		var (
+			chunkSize uint64 = 1000
+			maxSize   uint64 = 1 << 20
+		)
+
+		for _, p := range peers {
+			wg.Add(1)
+			go func(conn *grpc.ClientConn) {
+				defer wg.Done()
+
+				ctx, cancel := context.WithTimeout(context.Background(), conf.GetDownloadTxTimeout())
+				defer cancel()
+
+				stream, err := NewWaveletClient(conn).SyncTransactions(ctx)
+				if err != nil {
+					logger.Error().Err(err).Msg("failed to create sync transactions stream")
+					return
+				}
+
+				defer func() {
+					if err := stream.CloseSend(); err != nil {
+						logger.Error().Err(err).Msg("failed to send sync transactions bloom filter")
+					}
+				}()
+
+				if err := stream.Send(bfReq); err != nil {
+					logger.Error().Err(err).Msg("failed to send sync transactions bloom filter")
+					return
+				}
 
-						responseChan <- nil
+				res, err := stream.Recv()
+				if err != nil {
+					logger.Error().Err(err).Msg("failed to receive sync transactions header")
+					return
+				}
+
+				count := res.GetTransactionsNum()
+				if count == 0 || count > maxSize {
+					return
```

Comment:
What do you think about logging the case where too many transactions are received (and thus we don't accept any?)

```suggestion
					if count != 0 {
						logger.Debug().Msg("Bad number of transactions would be received").Uint64("count", count)
					}
					return
```